### PR TITLE
Feature/EPMEDU-594 View nearby feeding points on map

### PR DESCRIPTION
**Pull request by ozkarram**
[Ticket link for reference](https://jira.epam.com/jira/browse/EPMEDU-594)

- Use case 1 - Open the Home page without a GPS permissions

https://github.com/AnimealProject/animeal_android/assets/4513325/87e80832-4116-434d-826e-c407a16c97c3

- Use case 2 - Open the Home page with a GPS permissions and "Red" feeding point nearby
  - Priority 1: One closest favorite

https://github.com/AnimealProject/animeal_android/assets/4513325/ff365dad-d8d6-403c-8d2d-b8d00c43d14b

  - Priority 2: One remote (more than 10 km) favorite

https://github.com/AnimealProject/animeal_android/assets/4513325/eabd0a53-8c3d-43be-8c0f-fb6629f02f22

  - Priority 3: Nearest (not favorite), but no more than in 10 kilometers

https://github.com/AnimealProject/animeal_android/assets/4513325/f3deb364-13e8-4361-bb2f-a210cbc8bc0c

- Use case 3 - Open the Home page with a GPS permissions if there are no "Red" feeding point nearby

https://github.com/AnimealProject/animeal_android/assets/4513325/040dea44-72ac-470b-b974-eebbfa5e2fdf

- Use case 4 - See the feeding points on the map

https://github.com/AnimealProject/animeal_android/assets/4513325/573706aa-a787-4487-aa4e-d22f629b5f51

**Description**
Uses Turf measurement lib to get distance between user's location and feeding points and select it according to priority rules.

### CODE REVIEW

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+class GetFeedingPointByPriorityUseCase {
+
+    /*
+     * Filters iterable of FeedingPointModel by priority defined at EPMEDU-594:
+     * - 1st: Status = Red, is favorite and is near to user location
+     * - 2nd: Status = Red, is favorite
+     * - 3rd: Status = Red, is not favorite but is near to user location
+     */
+    operator fun invoke(
+        feedingPoints: ImmutableList<FeedingPointModel>,
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Can we instead get feeding points from `FeedingPointRepository`, because we should use `FeedingPoint` in `domain` layer, but not `FeedingPointModel`?

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        userLocation: MapLocation
+    ): FeedingPointModel? {
+        val readyToFeed = feedingPoints.filter { it.feedStatus == FeedStatus.RED }
+        return if (readyToFeed.isNotEmpty()) {
+            readyToFeed.firstOrNull { it.isFavourite && it.isNearTo(userLocation) } ?: run {
+                readyToFeed.firstOrNull { it.isFavourite }
+            } ?: readyToFeed.firstOrNull { it.isNearTo(userLocation) }
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: We must return the nearest feeding point, not the first in the list which is near

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedingPointModel.kt [hunk 3]
```diff
@@ -64,4 +66,10 @@ data class FeedingPointModel(
                 }
             }
         }
+
+    companion object {
+        private const val NEAREST_DISTANCE_KM = 10
+        fun FeedingPointModel.isNearTo(location: MapLocation): Boolean =
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedingPointModel.kt]: Let's move this function to location module and detach it from feeding points

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt [hunk 4]
```diff
@@ -111,6 +114,12 @@ class DefaultFeedingPointHandler @Inject constructor(
         }
     }
 
+    override fun CoroutineScope.fetchNearestFeedingPoint(userLocation: MapLocation) {
+        val nearestPoint = getFeedingPointByPriorityUseCase(state.feedingPoints, userLocation)
+        nearestPoint?.let {
+            handleFeedingPointEvent(Select(nearestPoint))
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt]: Use `selectFeedingPoint` function directly

**Review by Fileek — CHANGES_REQUESTED** [review]

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt (hunk 2 above)
Comment (ozkarram) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Comparator added.

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedingPointModel.kt (hunk 3 above)
Comment (ozkarram) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedingPointModel.kt]: Moved out to extension function at library > geolocation package

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 5]
```diff
@@ -1,30 +1,53 @@
...
-        return if (readyToFeed.isNotEmpty()) {
-            readyToFeed.firstOrNull { it.isFavourite && it.isNearTo(userLocation) } ?: run {
-                readyToFeed.firstOrNull { it.isFavourite }
-            } ?: readyToFeed.firstOrNull { it.isNearTo(userLocation) }
-        } else {
-            null
+    suspend operator fun invoke(
+        type: AnimalType,
+        userLocation: MapLocation,
+        onFeedingPointFound: (FeedingPointModel) -> Unit,
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Please use `suspendCancellableCoroutine` instead of the callback. You can find examples in `AnimealApi` or `AuthApi`

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    /*
+     * Filters iterable of FeedingPointModel by priority defined at EPMEDU-594:
+     * - 1st: Status = Red, is favorite and is near to user location
+     * - 2nd: Status = Red, is favorite
+     * - 3rd: Status = Red, is not favorite but is near to user location
+     */
+    suspend operator fun invoke(
+        type: AnimalType,
+        userLocation: MapLocation,
+        onFeedingPointFound: (FeedingPointModel) -> Unit,
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Domain layer should not know about presentation layer. Return `FeedingPoint`, not `FeedingPointModel`

**Review by Fileek — CHANGES_REQUESTED** [review]

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    ) {
+        getAllFeedingPointsUseCase(type = type).collect { domainFeedingPoints ->
+            val readyToFeed = domainFeedingPoints
+                .filter { it.animalStatus == AnimalState.RED }
```
Comment (VEZE) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: We have repository with getFeedingPointsBy(predicate) method, better to modify getAllFeedingPointsUseCase and add support for alternate predicate, or just implement repo method with specified predicates

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                .sortByNearestToLocation(userLocation)
+            if (readyToFeed.isNotEmpty()) {
+                val feedPointFound = readyToFeed
+                    .firstOrNull {
```
Comment (VEZE) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Can use find, this is alias for firstOrNull

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                        it.isFavourite && it.location.toPoint().isNearTo(userLocation.toPoint())
```
Comment (VEZE) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Why we need those logic? wouldn't it be easier to sort the list by multiple predicates, according to priority?

File: library/geolocation/src/main/java/com/epmedu/animeal/geolocation/location/Ext.kt [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.epmedu.animeal.geolocation.location
+
+import com.mapbox.geojson.Point
+import com.mapbox.turf.TurfMeasurement
+
+private const val NEAREST_DISTANCE_KM = 10
```
Comment (VEZE) [inline library/geolocation/src/main/java/com/epmedu/animeal/geolocation/location/Ext.kt]: Rename to default_nearest_distance_km, smth simillar
Not sure about private modifier

File: feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt [hunk 11]
```diff
@@ -141,15 +148,19 @@ internal class HomeViewModel @Inject constructor(
...
+                stateFlow
+            ) { feedingPointUpdate, feedingUpdate, feedingRouteUpdate, homeStateFlow ->
                 updateState {
                     copy(
                         feedingPointState = feedingPointUpdate,
                         feedState = feedingUpdate,
                         feedingRouteState = feedingRouteUpdate,
                     )
                 }
+                if (feedingPointUpdate.feedingPoints.isNotEmpty() && !homeStateFlow.locationState.isUndefined) {
```
Comment (VEZE) [inline feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt]: Can you also refactor isUndefined logic within the current task? Remove underlying val's overriding

**Review by VEZE — CHANGES_REQUESTED** [review]

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt (hunk 9 above)
Comment (ozkarram) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Makes sense from the readability point of view. However a predicate for each priority would iterate completely the whole set of feeding points (internally `findOrNull` and `find` is a `for` cycle with an `if()`) which could increase the amount of processing.

File: library/geolocation/src/main/java/com/epmedu/animeal/geolocation/location/Ext.kt (hunk 10 above)
Comment (ozkarram) [reply library/geolocation/src/main/java/com/epmedu/animeal/geolocation/location/Ext.kt]: Renamed, not needed at any other place.

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt (hunk 5 above)
Comment (ozkarram) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Returned as `Flow<FeedingPoint>` as we talked offline

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt (hunk 9 above)
Comment (VEZE) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Didn't catch your point, talking about those
Varargs predicates
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/compare-by.html

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt (hunk 7 above)
Comment (VEZE) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: not resolved

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt [hunk 12]
```diff
@@ -111,6 +114,16 @@ class DefaultFeedingPointHandler @Inject constructor(
         }
     }
 
+    override fun CoroutineScope.fetchNearestFeedingPoint(userLocation: MapLocation) {
+        launch {
+            getFeedingPointByPriorityUseCase(state.defaultAnimalType, userLocation)
+                .collect { result ->
+                    result?.let {
+                        selectFeedingPoint(FeedingPointModel(it))
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt]: Let's cancel this collect to avoid multiple selecting

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                val (favourite, notFavourite) = feedingPoints.partition { it.isFavourite }
+                val priority = when {
+                    favourite.isNotEmpty() -> Favourite(favourite) // 1st and 2nd priority
+                    else -> NotFavourite(notFavourite) // 3rd priority
+                }
+                return priority.getPrioritisedFeedingPoints(userLocation).firstOrNull()
+            }
+        }
+    }
+}
```
Comment (VEZE) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt:66]:
```suggestion
package com.epmedu.animeal.feeding.domain.usecase

import com.epmedu.animeal.feeding.domain.model.FeedingPoint
import com.epmedu.animeal.feeding.domain.model.enum.AnimalState
import com.epmedu.animeal.feeding.domain.repository.FeedingPointRepository
import com.epmedu.animeal.feeding.presentation.model.MapLocation
import com.epmedu.animeal.feeding.presentation.model.MapLocation.Companion.toPoint
import com.epmedu.animeal.foundation.tabs.model.AnimalType
import com.epmedu.animeal.geolocation.location.distanceInKmTo
import com.epmedu.animeal.geolocation.location.isNearTo
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.map

class GetFeedingPointByPriorityUseCase(
    private val feedingPointRepository: FeedingPointRepository
) {
    /*
     * Filters list of FeedingPoint by priority defined at EPMEDU-594:
     * - 1st: Status = Red, is favorite and is near to user location
     * - 2nd: Status = Red, is favorite
     * - 3rd: Status = Red, is not favorite but is near to user location
     */
    operator fun invoke(type: AnimalType, userLocation: MapLocation): Flow<FeedingPoint> {
        return feedingPointRepository.getFeedingPointsBy(shouldFetch = false) {
            it.animalType == type && it.animalStatus == AnimalState.RED
        }.map { feedingPoints ->
            FeedingPointsPriority.Factory(feedingPoints).resolvePriority(userLocation)
        }.filterNotNull()
    }

    sealed interface FeedingPointsPriority {

        val feedingPoints: List<FeedingPoint>

        fun getPrioritisedFeedingPoints(userLocation: MapLocation): List<FeedingPoint>

        class Favourite(override val feedingPoints: List<FeedingPoint>) : FeedingPointsPriority {

            override fun getPrioritisedFeedingPoints(userLocation: MapLocation): List<FeedingPoint> =
                feedingPoints.sortedWith(compareBy {
                    it.location.toPoint().distanceInKmTo(userLocation.toPoint())
                })
        }

        class NotFavourite(override val feedingPoints: List<FeedingPoint>) : FeedingPointsPriority {

            override fun getPrioritisedFeedingPoints(userLocation: MapLocation): List<FeedingPoint> =
                feedingPoints.filter {
                        it.location.toPoint().isNearTo(userLocation.toPoint())
                    }.sortedWith(compareBy {
                        it.location.toPoint().distanceInKmTo(userLocation.toPoint())
                    })
        }

        class Factory(private val feedingPoints: List<FeedingPoint>) {

            fun resolvePriority(userLocation: MapLocation): FeedingPoint? {
                val (favourite, notFavourite) = feedingPoints.partition { it.isFavourite }
                val priority = when {
                    favourite.isNotEmpty() -> Favourite(favourite) // 1st and 2nd priority
                    else -> NotFavourite(notFavourite) // 3rd priority
                }
                return priority.getPrioritisedFeedingPoints(userLocation).firstOrNull()
            }
        }
    }
}

```

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt [hunk 14]
```diff
@@ -115,7 +116,8 @@ class DefaultFeedingPointHandler @Inject constructor(
     }
 
     override fun CoroutineScope.fetchNearestFeedingPoint(userLocation: MapLocation) {
-        launch {
+        fetchByPriorityJob?.cancel()
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt:119]: I meant canceling the coroutine inside right after selecting, this line doesn't prevent multiple selecting

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    /*
+     * Filters list of FeedingPoint by priority defined at EPMEDU-594:
+     * - 1st: Status = Red, is favorite and is near to user location
+     * - 2nd: Status = Red, is favorite
+     * - 3rd: Status = Red, is not favorite but is near to user location
+     */
+    operator fun invoke(
+        type: AnimalType,
+        userLocation: MapLocation,
+    ): Flow<FeedingPoint?> {
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Let's return not nullable `FeedingPoint` to avoid using `let` by adding `filterNotNull` into the chain below

File: feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt [hunk 16]
```diff
@@ -142,15 +148,21 @@ internal class HomeViewModel @Inject constructor(
...
+                stateFlow
+            ) { feedingPointUpdate, feedingUpdate, feedingRouteUpdate, homeStateFlow ->
                 updateState {
                     copy(
                         feedingPointState = feedingPointUpdate,
                         feedState = feedingUpdate,
                         feedingRouteState = feedingRouteUpdate,
                     )
                 }
+                if (feedingPointUpdate.feedingPoints.isNotEmpty() &&
```
Comment (Fileek) [inline feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt]: Can we move this first check into the usecase?

Comment (ozkarram) [reply feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt]: Moved to usecase

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt (hunk 14 above)
Comment (ozkarram) [reply shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/viewmodel/handler/feedingpoint/DefaultFeedingPointHandler.kt:119]: `take(1)` at usecase prevents this

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt [hunk 17]
```diff
@@ -24,43 +26,48 @@ class GetFeedingPointByPriorityUseCase(
...
                     }
                 )
         }
+
         class NotFavourite(override val feedingPoints: List<FeedingPoint>) : FeedingPointsPriority {
+
             override fun getPrioritisedFeedingPoints(userLocation: MapLocation): List<FeedingPoint> =
                 feedingPoints.filter {
                     it.location.toPoint().isNearTo(userLocation.toPoint())
-                }.sortedWith(compareBy { it.location.toPoint().distanceInKmTo(userLocation.toPoint()) })
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt]: Why did you remove sorting here?

**Review by Fileek — APPROVED** [review]

**Review by VEZE — APPROVED** [review]
